examples/secret_storage/secret_storage.go

```diff
@@ -0,0 +1,572 @@
+	result, err = r.Lookup(teamName, tool)
+	if err != nil {
+		return false, result, err // api call failed
+	} else if result.EntryValue == "" {
+		return true, result, nil // tool already doesn't exist
+	}
+
+	expectedRevision := result.Revision + 1
+	_, err = r.api.DeleteEntryWithRevision(teamName, r.namespace, tool, expectedRevision)
+	if err != nil {
+		if _, ok := err.(kbchat.Error); !ok {
+			// unexpected error
+			return false, result, err
+		}
+
+		// failed delete. try get
+		result, err := r.Lookup(teamName, tool)
+		if err != nil {
+			return false, result, err // api call failed
+		}
+		return false, result, nil // failed delete. return KVGetResult
+	}
+	return true, result, nil // successul delete
+}
+
+// Reserve reserve a tool for a given day if that day is already not reserved.
+// Note: if you reserve a not-added or deleted tool, it will add the tool.
+// Returns (whether action is successful, most recent get result if applicable, error)
+func (r *RentalBotClient) Reserve(teamName string, username string, tool string, day string) (ok bool, result keybase1.KVGetResult, err error) {
+	var val map[string]string
+	result, err = r.Lookup(teamName, tool)
+	if err != nil {
+		return false, result, err // api call failed
+	} else if result.EntryValue == "" {
+		val = make(map[string]string)
+	} else {
+		if err = json.Unmarshal([]byte(result.EntryValue), &val); err != nil {
+			return false, result, err
+		}
+	}
+
+	if _, ok := val[day]; ok {
+		return false, result, nil // failed to put because day is already reserved
+	}
+	val[day] = username
+	expectedRevision := result.Revision + 1
+
+	bytes, err := json.Marshal(val)
+	if err != nil {
+		return false, result, err
+	}
+	_, err = r.api.PutEntryWithRevision(teamName, r.namespace, tool, string(bytes), expectedRevision)
+	if err != nil {
+		if e, ok := err.(kbchat.Error); !ok || e.Code != kbchat.RevisionErrorCode {
+			// unexpected error
+			return false, result, err
+		}
+
+		// failed put. try get
+		result, err := r.Lookup(teamName, tool)
+		if err != nil {
+			return false, result, err // api call failed
+		}
+		return false, result, nil // failed put. return KVGetResult
+	}
+	return true, result, nil // successul put
+}
+
+// Unreserve a tool for a given day if that day is currently reserved by the given user.
+// Note: if you unreserve a not-added or deleted tool, it will not add the tool.
+// Returns (whether action is successful, most recent get result if applicable, error)
+func (r *RentalBotClient) Unreserve(teamName string, username string, tool string, day string) (ok bool, result keybase1.KVGetResult, err error) {
+	var val map[string]string
+	result, err = r.Lookup(teamName, tool)
+	if err != nil {
+		return false, result, err // api call failed
+	} else if result.EntryValue == "" {
+		val = make(map[string]string)
+	} else {
+		if err := json.Unmarshal([]byte(result.EntryValue), &val); err != nil {
+			return false, result, err
+		}
+	}
```

Comment:
any chance we can dry out some of this code with the `Reserve` method? maybe something like `fetchReservations`?